File: src/client/AM.PA.MonitoringTool/WindowsActivityTracker/Visualizations/DayMostFocusedProgram.cs
// Created by André Meyer (ameyer@ifi.uzh.ch) from the University of Zurich
// Created: 2015-12-09
// 
// Licensed under the MIT License.

using Shared;
using Shared.Helpers;
using System;
using System.Globalization;
using WindowsActivityTracker.Data;

namespace WindowsActivityTracker.Visualizations
{
    internal class DayMostFocusedProgram : BaseVisualization, IVisualization
    {
        private readonly DateTimeOffset _date;
        private const int _minFocusTime = 2;

        public DayMostFocusedProgram(DateTimeOffset date)
        {
            this._date = date;

            Title = "Longest Time<br />Focused in a Program";
            IsEnabled = true; //todo: handle by user
            Order = 20; //todo: handle by user
            Size = VisSize.Small;
            Type = VisType.Day;
        }

        public override string GetHtml()
        {
            var html = string.Empty;

            /////////////////////
            // fetch data sets
            /////////////////////
            var queryResultsLocal = Queries.GetLongestFocusOnProgram(_date);
            var durInMin = (queryResultsLocal == null) ? 0 : queryResultsLocal.DurationInSec / 60.0;

            if (queryResultsLocal == null || durInMin <= _minFocusTime)
            {
                html += VisHelper.NotEnoughData(string.Format(CultureInfo.InvariantCulture, "We either don't have enough data or you didn't focus on a single program for more than {0} minutes on this day.", _minFocusTime));
                return html;
            }

            /////////////////////
            // HTML
            /////////////////////
            html += "<p style='text-align: center; margin-top:-0.7em;'><strong style='font-size:2.5em; color:" + Shared.Settings.RetrospectionColorHex + ";'>" + Math.Round(durInMin, 0) + "</strong>min</p>";
            html += string.Format(CultureInfo.InvariantCulture, "<p style='text-align: center; margin-top:-0.7em;'>in {0}<br />from {1} to {2}</p>",
                ProcessNameHelper.GetFileDescription(queryResultsLocal.Process),
                queryResultsLocal.From.ToShortTimeString(),
                queryResultsLocal.To.ToShortTimeString());

            return html;
        }
    }
}

File: src/client/AM.PA.MacOs/PersonalAnalytics/Trackers/Activity Tracker/Visualizations/DayMostFocusedProgram.swift
//
//  DayFragmentationTimeline.swift
//  PersonalAnalytics
//
//  Created by Chris Satterfield on 2017-05-29.
//
//  Adapted from Windows version created by André Meyer


import Foundation


class DayMostFocusedProgram: Visualization{
    
    let Size: String
    let color = AppConstants.retrospectiveColor
    let title: String
    let sql: ActivitySQLController
    var _type: [String] = [VisConstants.Day]
    let _minFocusTime = 2.0

    
    required init() throws {
        Size = "Small"
        title = "Longest Time Focused in an App"
        sql = try ActivitySQLController()
    }
    
    struct FocusedWorkDict{
        var name: String
        var to: TimeInterval
        var from: TimeInterval
        var difference: TimeInterval
        
        init(name: String, to: TimeInterval, from: TimeInterval, difference: TimeInterval){
            self.name = name
            self.to = to
            self.from = from
            self.difference = difference
        }
        
    }
    
    func getHtml(_ _date: Date, type: String) -> String {
        
        if(!_type.contains(type)){
            return ""
        }
        var html = ""
        
        /////////////////////
        // fetch data sets
        /////////////////////
        let queryResultsLocal = sql.GetLongestFocusOnProgram(date: _date);
        
        if(queryResultsLocal == nil){
            return ""
        }
        
        let durInMin = (queryResultsLocal == nil) ? 0 : queryResultsLocal!.difference / 60.0;
        
        let from = Date(timeIntervalSince1970: queryResultsLocal!.from)
        let to = Date(timeIntervalSince1970: queryResultsLocal!.to)
        
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "h:mma"

        
        
        
        if (queryResultsLocal == nil || durInMin <= _minFocusTime)
        {
            html += VisHelper.NotEnoughData("We either don't have enough data or you didn't focus on a single program for more than \(_minFocusTime) minutes on this day.")
            return html;
        }
        
        /////////////////////
        // HTML
        /////////////////////
        html += "<p style='text-align: center; margin-top:-0.7em;'><strong style='font-size:2.5em; color:" + color + ";'>" + String(durInMin.rounded()) + "</strong> min</p>"
        html += "<p style='text-align: center; margin-top:-0.7em;'>in \(queryResultsLocal!.name) <br />from \(dateFormatter.string(from: from)) to \(dateFormatter.string(from: to))</p>"
        
        return html
    }

}
